librad/src/identities/git.rs

```diff
@@ -0,0 +1,763 @@
+            {
+                Ok(Action::Uptodate)
+            } else if theirs.signatures.contains_key(&our_pk)
+                && self
+                    .repo
+                    .graph_descendant_of(*theirs.content_id, *ours.content_id)?
+            {
+                Ok(Action::FastFwd)
+            } else if ours.revision == theirs.revision {
+                Ok(Action::SlowFwd)
+            } else if Some(&ours.revision) == theirs.doc.replaces() {
+                Ok(Action::SuccRev)
+            } else {
+                Err(error::Merge::RevisionMismatch)
+            }
+        }?;
+
+        //println!("action: {:?}", action);
+
+        match action {
+            Action::Uptodate => Ok(ours),
+            Action::FastFwd => Ok(theirs),
+            Action::SlowFwd => {
+                let mut signatures = ours.signatures.clone();
+                signatures.extend(theirs.signatures.clone());
+
+                let content_id = self.commit(
+                    &format!("Updated signatures from {}", theirs.content_id),
+                    &signatures,
+                    ours.revision,
+                    &[&ours],
+                )?;
+
+                Ok(Identity {
+                    content_id,
+                    signatures,
+                    ..ours
+                })
+            },
+            Action::SuccRev => {
+                let mut signatures = theirs.signatures.clone();
+                {
+                    let sig = sign(signer, theirs.revision).map_err(error::Merge::Signer)?;
+                    signatures.extend(Some(sig))
+                }
+
+                let content_id = self.commit(
+                    &format!(
+                        "Approved new revision `{}` from {}",
+                        theirs.revision, theirs.content_id
+                    ),
+                    &signatures,
+                    theirs.revision,
+                    &[&ours, &theirs],
+                )?;
+
+                Ok(Identity {
+                    content_id,
+                    signatures,
+                    ..theirs
+                })
+            },
+        }
+    }
+
+    //// Helpers ////
+
+    fn commit(
+        &self,
+        message: &str,
+        signatures: &Signatures,
+        revision: Revision,
+        parents: &[&Identity<T>],
+    ) -> Result<ContentId, git2::Error> {
+        let tree = self.repo.find_tree(*revision)?;
+        let parents = parents
+            .iter()
+            .map(|parent| self.repo.find_commit(*parent.content_id))
+            .collect::<Result<Vec<_>, _>>()?;
+        let author = self.repo.signature()?;
+
+        self.repo
+            .commit(
+                None,
+                &author,
+                &author,
+                &sign::CommitMessage::new(message, signatures).to_string(),
+                &tree,
+                parents.iter().collect::<Vec<_>>().as_slice(),
```

Comment:
Since you collected above couldn't you just call `as_slice`? Or is it something to do with references? :roll_eyes: 
```suggestion
                parents.as_slice(),
```

Reply:
Oh ja, very annoying: it wants a `&[&Commit<'_>]`. Dunno how to make one without allocating two `Vec`s.

Reply:
Fair enough :(